New scale-out and redis backplane setup docs (#9556)
aspnet/Docs: aspnetcore/signalr/scale.md

---
title: ASP.NET Core SignalR production hosting and scaling
author: tdykstra
description: Learn how to avoid performance and scaling problems in apps that use ASP.NET Core SignalR.
monikerRange: '>= aspnetcore-2.1'
ms.author: tdykstra
ms.custom: mvc
ms.date: 11/28/2018
uid: signalr/scale
---

# ASP.NET Core SignalR hosting and scaling

By [Andrew Stanton-Nurse](https://twitter.com/anurse), [Brady Gaster](https://twitter.com/bradygaster), and [Tom Dykstra](https://github.com/tdykstra),

This article explains hosting and scaling considerations for high-traffic apps that use ASP.NET Core SignalR.

## TCP connection resources

> The number of concurrent TCP connections that a web server can support is limited. Standard HTTP requests use these limited resources efficiently, opening and closing a connection for every request and response. A SignalR connection holds the TCP connection open for a much longer time. In a high-traffic app that serves many clients, these persistent connections can cause servers to hit their maximum number of connections.

Reviewer: ```suggestion
The number of concurrent TCP connections that a web server can support is limited. Standard HTTP clients use *ephemeral* connections. These connections can be closed when the client goes idle and re-opened later. On the other hand, a SignalR connection is *persistent*. SignalR connections stay open even when the client goes idle. In a high-traffic app that serves many clients, these persistent connections can cause servers to hit their maximum number of connections.
```

Had another thought about rephrasing this. We also want to avoid the nit-pickers who say that HTTP requests do use connections "efficiently" and don't just keep opening and closing connections ;).

Persistent connections also consume some additional memory, to track each connection.

The heavy use of connection-related resources by SignalR can affect other web apps that are hosted on the same server. When SignalR opens and holds the last available TCP connections, other web apps on the same server also have no more connections available to them.

> If a server runs out of connections, you'll see random socket errors and connection reset errors, for example:

Reviewer 2: ```suggestion
If a server runs out of connections, you'll see random socket errors and connection reset errors. For example:
```

```
An attempt was made to access a socket in a way forbidden by its access permissions...
```

> To keep SignalR resource usage from causing errors in other web apps, run SignalR on different servers than your other web apps.

Reviewer 2: ```suggestion
To prevent SignalR resource consumption from causing errors in other web apps, run SignalR on different servers than your other web apps.
```

> To keep SignalR resource usage from causing errors in a SignalR app, scale out to limit the number of connections a server has to handle.

Reviewer 2: ```suggestion
To prevent SignalR resource consumption from causing errors in a SignalR app, scale out to limit the number of connections a server has to handle.
```

## Scale out

> Unlike a stateless web app, you can't scale out a SignalR app simply by increasing the number of servers to handle increased traffic. New servers get new connections, and SignalR on server 2 knows nothing about the connections on servers 1 and 3. When SignalR on server 2 wants to send a message to all clients, the message only goes to the clients connected to that server. The following diagram illustrates this scenario.

Reviewer 3: Unlike a stateless web app, SignalR can't be scaled out ~simply~ by increasing the number of servers to handle increased traffic.

Reviewer 3: I like to avoid you unless it helps clarity.

![Scaling SignalR without a backplane](scale/_static/scale-no-backplane.png)

> There are two recommended solutions to this problem:

Reviewer 3: Maybe something like

Azure SignalR Service or the Redis backplane can be used to scale SignalR.

> * Azure SignalR Service

Author: I made these internal links to the the respective h2 sections.  The "more info" links are at the end of the SignalR Service section and at the end of the page.

* Redis backplane

> 

Reviewer 2: Delete extra blank line

## Azure SignalR Service

The Azure SignalR Service is a proxy rather than a backplane. Each time a client initiates a connection to the server, the client is redirected to connect to the service. That process is illustrated in the following diagram:

![Establishing a connection to the Azure SignalR Service](scale/_static/azure-signalr-service-one-connection.png)

The result is that the service manages all of the client connections, while each server needs only a small constant number of connections to the service, as shown in the following diagram:

![Clients connected to the service, servers connected to the service](scale/_static/azure-signalr-service-multiple-connections.png)

This approach to scale-out has several advantages over the Redis backplane alternative:

> * The SignalR app can scale out without requiring sticky sessions, because clients are immediately redirected to the Azure service when they connect.

Reviewer 3: sticky sessions should be a link if possible.

> * The SignalR app can scale out without requiring sticky sessions, because clients are immediately redirected to the Azure service when they connect.

Reviewer 3: The SignalR app should be `The SignalR service`. Since that's repeated mabe something like:

The SignalR service has several scale-out advantages over the Redis backplane. The SignalR service:

* can scale out without requiring sticky sessions, because clients are immediately redirected to the Azure service when they connect.
* can scale out based on the number of messages sent, while the Azure service automatically scales to handle any number of connections. For example, there could be thousands of clients, but if only a few messages per second are sent, the SignalR app won't need to scale out to multiple servers just to handle the connections themselves.
* won't use significantly more connection resources than a web app without SignalR.

Author: The bullets are about an app that uses SignalR, not about the SignalR service -- I'm rewriting to make that clearer.

> * The SignalR app can scale out based on the number of messages sent, while the Azure service automatically scales to handle any number of connections. For example, there could be thousands of clients, but if only a few messages per second are sent, the SignalR app won't need to scale out to multiple servers just to handle the connections themselves.

Reviewer 2: ```suggestion
* The SignalR app can scale out based on the number of messages sent, while the Azure service automatically scales to handle any number of connections. For example, there could be thousands of clients. If only a few messages per second are sent, the SignalR app won't need to scale out to multiple servers just to handle the connections themselves.
```

* The SignalR app won't use significantly more connection resources than a web app without SignalR.

> For these reasons, we recommend the Azure SignalR Service for all ASP.NET Core SignalR apps that run on Azure, including App Service, VMs, and containers.

Reviewer 2: ```suggestion
For these reasons, we recommend the Azure SignalR Service for all ASP.NET Core SignalR apps hosted on Azure, including App Service, VMs, and containers.
```

> For these reasons, we recommend the Azure SignalR Service for all ASP.NET Core SignalR apps that run on Azure, including App Service, VMs, and containers.

Author: I'm adding it to the TOC now, will create an issue for converting it to "contextual ToC" to avoid switching to the Azure ToC when you click on the link in the ASP.NET ToC.

> For more information see the [Azure SignalR Service documentation](https://docs.microsoft.com/en-us/azure/azure-signalr/signalr-overview).

Reviewer 2: ```suggestion
For more information see the [Azure SignalR Service documentation](/azure/azure-signalr/signalr-overview).
```

## Redis backplane

> Redis is an in-memory key-value store that supports a messaging system with a publish/subscribe model. The SignalR Redis backplane uses the pub/sub feature to forward messages to other servers. When a client makes a connection, the connection information is passed to the backplane. Then when a server wants to send a message to all clients, it sends to the backplane. The backplane knows all connected clients and which servers they're on. It sends the message to all clients via their respective servers. This process is illustrated in the following diagram:

Reviewer 3: ~Then~ when a server wants to send a message to all clients,

> Redis is an in-memory key-value store that supports a messaging system with a publish/subscribe model. The SignalR Redis backplane uses the pub/sub feature to forward messages to other servers. When a client makes a connection, the connection information is passed to the backplane. Then when a server wants to send a message to all clients, it sends to the backplane. The backplane knows all connected clients and which servers they're on. It sends the message to all clients via their respective servers. This process is illustrated in the following diagram:

Reviewer 2: ```suggestion
Redis is an in-memory key-value store that supports a messaging system with a publish/subscribe (pub/sub) model. The SignalR Redis backplane uses the pub/sub feature to forward messages to other servers. When a client makes a connection, the connection information is passed to the backplane. When a server wants to send a message to all clients, it sends to the backplane. The backplane knows all connected clients and which servers they're on. It sends the message to all clients via their respective servers. This process is illustrated in the following diagram:
```

![Redis backplane, message sent from one server to all clients](scale/_static/redis-backplane.png)

> We recommend the Redis backplane as the scale-out approach for apps that run on your own infrastructure. Azure SignalR Service is not a practical option for production use with on-premises apps due to latency of a connection between your data center and an Azure data center.

Reviewer 2: ```suggestion
The Redis backplane is the recommended scale-out approach for apps hosted on your own infrastructure. Azure SignalR Service isn't a practical option for production use with on-premises apps due to connection latency between your data center and an Azure data center.
```
